File: farondroid/src/main/java/com/openfarmanager/android/filesystem/FileSystemScanner.java
package com.openfarmanager.android.filesystem;

import android.os.Build;
import android.text.TextUtils;
import com.openfarmanager.android.App;
import com.openfarmanager.android.filesystem.actions.RootTask;
import com.openfarmanager.android.filesystem.filter.Sorter;
import com.openfarmanager.android.filesystem.filter.SorterFactory;
import com.openfarmanager.android.model.exeptions.FileIsNotDirectoryException;
import com.openfarmanager.android.utils.StorageUtils;

import org.apache.commons.io.IOCase;
import org.apache.commons.io.filefilter.PrefixFileFilter;

import java.io.File;
import java.util.*;

/**
 * Vlad Namashko
 */
public class FileSystemScanner {

    public static String ROOT = "/";

    public static final FileSystemScanner sInstance;

    private LinkedList<Sorter> mSorters;

    static {
        sInstance = new FileSystemScanner();
    }

    private Comparator<FileProxy> mComparator = new Comparator<FileProxy>() {
        public int compare(FileProxy f1, FileProxy f2) {
            int result = 0;
            for (Sorter sorter : mSorters) {
                if ((result = sorter.doSort(f1, f2)) != 0) {
                    return result;
                }
            }
            return result;
        }
    };

    private FileSystemScanner() {
        initSorters();
    }

    public void initSorters() {
        ROOT = App.sInstance.getSettings().isSDCardRoot() ? StorageUtils.getSdPath() : "/";

        mSorters = new LinkedList<>();
        if (App.sInstance.getSettings().isFoldersFirst()) {
            mSorters.add(SorterFactory.createDirectoryUpFilter());
        }
        mSorters.add(SorterFactory.createPreferredFilter());
    }

    public static Collection<File> getTree(File... root) {
        ArrayList<File> tree = new ArrayList<>();
        for (File f : root) {
            addFilesRecursively(f, tree);
        }
        return tree;
    }

    private static void addFilesRecursively(File file, Collection<File> all) {
        final File[] children = file.listFiles();
        if (children != null) {
            for (File child : children) {
                all.add(child);
                addFilesRecursively(child, all);
            }
        }
    }

    public File getRoot() {
        return new File(ROOT);
    }

    public boolean isRoot(File node) {
        return node.getAbsolutePath().equals(ROOT);
    }

    public List<FileProxy> fallingDown(File currentNode, String mFilter) throws FileIsNotDirectoryException {
        if (!currentNode.isFile()) {
            String[] files = null;

            if (Build.VERSION.SDK_INT >= 24 && StorageUtils.getSdPath().startsWith(currentNode.getAbsolutePath())
                    && !StorageUtils.getSdPath().equals(currentNode.getAbsolutePath())) {

                files = RootTask.ls(currentNode);
                if (files != null) {
                    return buildFileList(currentNode, files, new LinkedList<>());
                }

                // don't allow to open subfolder of sd card root on SDK >= 24
                return null;
            }

            List<FileProxy> result = new LinkedList<FileProxy>();

            if (currentNode.canRead()) {
                if (!TextUtils.isEmpty(mFilter)) {
                    try {
                        files = currentNode.list(new PrefixFileFilter(mFilter, IOCase.INSENSITIVE));
                    } catch (Exception e) {
                        e.printStackTrace();
                    }
                }
                if (files == null) {
                    files = currentNode.list();
                }
            } else {
                files = RootTask.ls(currentNode);
            }

            if (files == null) {
                return Build.VERSION.SDK_INT >= 24 ? null : result;
            } else {
                return buildFileList(currentNode, files, result);
            }
        } else {
            throw new FileIsNotDirectoryException(currentNode.getAbsolutePath());
        }
    }

    private List<FileProxy> buildFileList(File currentNode, String[] files, List<FileProxy> result) {
        for (String f : files) {
            FileSystemFile file = new FileSystemFile(currentNode, f);
            if (App.sInstance.getSettings().isHideSystemFiles() && file.isHidden()) {
                continue;
            }
            result.add(file);
        }
        sort(result);
        return result;
    }

    public void sort(List<FileProxy> filesToSort) {
        Collections.sort(filesToSort, mComparator);
    }

}

File: farv2/src/main/java/com/openfarmanager/android/core/filesystem/FileSystemScanner.kt
package com.openfarmanager.android.core.filesystem

import android.os.Build
import com.openfarmanager.android.core.filesystem.exceptions.OpenDirectoryException
import com.openfarmanager.android.core.filesystem.exceptions.OpenDirectoryNotAllowed
import com.openfarmanager.android.model.Entity
import com.openfarmanager.android.model.FileEntity
import com.openfarmanager.android.model.UpNavigator
import com.openfarmanager.android.theme.ThemePref
import java.io.File
import java.util.*
import javax.inject.Inject

class FileSystemScanner @Inject constructor(
        val themePref: ThemePref,
        val storageUtils: StorageUtils) {

    private val sorters = LinkedList<Sorter>().apply {
        if (themePref.folderFirst) {
            add(SorterFactory.createDirectoryUpFilter())
        }
        add(SorterFactory.createPreferredFilter(themePref))
    }

    private val comparator = Comparator<Entity> { entity1, entity2 ->
        var result = 0
        sorters.forEach {sorter ->
            result = sorter.doSort(entity1, entity2)
            if (result != 0) {
                return@Comparator result
            }
        }
        result
    }

    fun openDirectory(directory: Entity, filter: String = ""): List<Entity> {
        if (directory.isDirectory) {

            if (Build.VERSION.SDK_INT >= 24
                    && storageUtils.sdPath.startsWith(directory.path)
                    && storageUtils.sdPath != directory.path) {

                // don't allow to open subfolder of sd card root on SDK >= 24
                throw OpenDirectoryNotAllowed("Not allowed to open subfolder of sd card root on SDK >= 24")
            }

            val file = File(directory.path)

            if (file.canRead()) {
                val list = mutableListOf<Entity>().apply {
                    add(UpNavigator(file.parentFile))

                    addAll(file.listFiles()
                            .map { item -> FileEntity(item) }
                            .filter { item -> !themePref.hideSystemFiles || !item.isHidden }
                            .toList()
                    )
                }
                sort(list)
                return list
            } else {
                throw OpenDirectoryNotAllowed("Can't read directory")
            }

        } else throw OpenDirectoryException("openDirectory can't scan file")
    }

    fun sort(filesToSort: List<Entity>) {
        Collections.sort(filesToSort, comparator)
    }
}